tools/checkrelease.sh

```diff
@@ -19,114 +19,221 @@
 #
 #############################################################################
 
-set -e
-
+VERBOSE=0
+RETURN_CODE=0
 BASE_URL="https://dist.apache.org/repos/dist/dev/incubator/nuttx"
-TEMPDIR="dist.apache.org"
+DIST_DIR="dist.apache.org"
+TMP="/tmp"
+TEMPDIR="$TMP/nuttx-checkrelease"
 ORIGINAL_DIR="$(pwd)"
-trap "{ cd $ORIGINAL_DIR; rm -rf $TEMPDIR; }" EXIT
+trap "{ rm -rf $TEMPDIR; }" EXIT
+
+function validate_url() {
+  if [[ `wget -S --spider $1  2>&1 | grep 'HTTP/1.1 200 OK'` ]]; then echo "true"; fi
+}
 
 function download_release() {
-    wget -r -np -R "index.html*" -P . --cut-dirs 7 "$URL"
-    cd "$TEMPDIR"
+  rm -rf "$TEMPDIR"
+  if [[ -n "$URL" ]]; then
+    mkdir "$TEMPDIR"
+    if [[ $(validate_url "$URL") ]]; then
+      echo "Downloading release files from $URL"
+      wget --quiet -r --no-parent -P "$TEMPDIR" --cut-dirs 100 "$URL"
+      cd "$TEMPDIR"
+      mv $DIST_DIR/apache-nuttx-* .
+    else
+      echo "The URL given doesn't return HTTP 200 OK return code— exiting. ($URL)"
+      exit 1
+    fi
+  else
+    if [[ -n "$DIRECTORY" ]]; then
+      cp -r "$DIRECTORY" "$TEMPDIR"
+      cd "$TEMPDIR"
+    else
+      echo "One of --dir or --url is required!"
+      exit 1
+    fi
+  fi
 }
 
 function check_sha512() {
-    # check release sha512
     RELEASE_FILE=$1
     echo "Checking $RELEASE_FILE sha512..."
-    sha512sum -c "$RELEASE_FILE.sha512"
+    output="$(sha512sum -c $RELEASE_FILE.sha512 2>&1)"
+    return_value=$?
+    if [ $return_value -eq 0 ]; then
+      echo " OK: $RELEASE_FILE sha512 hash matches."
+    else
+      RETURN_CODE=1
+      echo " - $RELEASE_FILE sha512 hash does not match:"
+      echo "$output"
+    fi
+    echo
 }
 
 function check_gpg() {
-    # check nuttx sha512 and gpg
     RELEASE_FILE=$1
     echo "Checking $RELEASE_FILE GPG signature:"
-    gpg --verify "$RELEASE_FILE.asc" "$RELEASE_FILE"
+    gpg --verify $RELEASE_FILE.asc $RELEASE_FILE
+    return_value=$?
+    if [ $return_value -eq 0 ]; then
+      echo " OK: $RELEASE_FILE gpg signature matches."
+    else
+      RETURN_CODE=1
+      echo " - Error checking $RELEASE_FILE gpg signature:"
+      echo "$output"
+      echo
```

Comment:
redundant?

Reply:
@xiaoxiang781216 The extra `echo` is there for formatting, to leave a blank line. I can take it out if you want, let me know.

Reply:
@xiaoxiang781216 re: wget output, here's mine:

```
$ tree
.

0 directories, 0 files
$ wget -q -r -nd -np https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1
$ tree
.
├── 9.1.0-RC1
├── apache-nuttx-9.0.0-incubating.tar.gz
├── apache-nuttx-9.0.0-incubating.tar.gz.1
├── apache-nuttx-9.0.0-incubating.tar.gz.asc
├── apache-nuttx-9.0.0-incubating.tar.gz.asc.1
├── apache-nuttx-9.0.0-incubating.tar.gz.sha512
├── apache-nuttx-9.0.0-incubating.tar.gz.sha512.1
├── apache-nuttx-9.1.0-incubating.tar.gz
├── apache-nuttx-9.1.0-incubating.tar.gz.1
├── apache-nuttx-9.1.0-incubating.tar.gz.2
├── apache-nuttx-9.1.0-incubating.tar.gz.asc
├── apache-nuttx-9.1.0-incubating.tar.gz.asc.1
├── apache-nuttx-9.1.0-incubating.tar.gz.asc.2
├── apache-nuttx-9.1.0-incubating.tar.gz.sha512
├── apache-nuttx-9.1.0-incubating.tar.gz.sha512.1
├── apache-nuttx-9.1.0-incubating.tar.gz.sha512.2
├── apache-nuttx-apps-9.0.0-incubating.tar.gz
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.1
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.asc
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.asc.1
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.sha512
├── apache-nuttx-apps-9.0.0-incubating.tar.gz.sha512.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.2
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc.2
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512.1
├── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512.2
├── index.html
├── index.html.1
├── index.html.2
├── index.html.3
├── index.html.4
├── KEYS
└── robots.txt

0 directories, 38 files
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
```

On my system that invocation downloads extra files. You can see from the first `tree` command above I'm starting with an empty directory.

Reply:
I hit the similar issue only when the temp directory isn't empty and wget rename the new download to *.1.

Reply:
@xiaoxiang781216 I've tried your invocation several times in empty directories, and get the same result as above. What temp directory are you referring to?

Reply:
@xiaoxiang781216 Here's the output from `wget --version`:

```
GNU Wget 1.19.4 built on linux-gnu.

-cares +digest -gpgme +https +ipv6 +iri +large-file -metalink +nls 
+ntlm +opie +psl +ssl/openssl 

Wgetrc: 
    /etc/wgetrc (system)
Locale: 
    /usr/share/locale 
Compile: 
    gcc -DHAVE_CONFIG_H -DSYSTEM_WGETRC="/etc/wgetrc" 
    -DLOCALEDIR="/usr/share/locale" -I. -I../../src -I../lib 
    -I../../lib -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_LIBSSL -DNDEBUG 
    -g -O2 -fdebug-prefix-map=/build/wget-Xb5Z7Y/wget-1.19.4=. 
    -fstack-protector-strong -Wformat -Werror=format-security 
    -DNO_SSLv2 -D_FILE_OFFSET_BITS=64 -g -Wall 
Link: 
    gcc -DHAVE_LIBSSL -DNDEBUG -g -O2 
    -fdebug-prefix-map=/build/wget-Xb5Z7Y/wget-1.19.4=. 
    -fstack-protector-strong -Wformat -Werror=format-security 
    -DNO_SSLv2 -D_FILE_OFFSET_BITS=64 -g -Wall -Wl,-Bsymbolic-functions 
    -Wl,-z,relro -Wl,-z,now -lpcre -luuid -lidn2 -lssl -lcrypto -lpsl 
    ftp-opie.o openssl.o http-ntlm.o ../lib/libgnu.a 

Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later
<http://www.gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Originally written by Hrvoje Niksic <hniksic@xemacs.org>.
Please send bug reports and questions to <bug-wget@gnu.org>.
```

What's your wget version?

Reply:
@xiaoxiang781216 aha. I figured it out. You need the trailing slash at the end of the URL. If I place it there, no duplicate files are created. I don't know why. I'll change the code to use your invocation and ensure there is a trailing slash at the end of the URL.